### PR TITLE
rpc: client_info: add retrieve_auxiliary_opt

### PR DESCRIPTION
**Pull request by bhalevy**
To facilitate optional passing of the host_id in client_info.

Signed-off-by: Benny Halevy <bhalevy@scylladb.com>

### CODE REVIEW

**Comment by bhalevy** [comment]
My plan is to use this in https://github.com/scylladb/scylladb/blob/5a07641b955c57c36cad0106127cfef96d0ee676/message/messaging_service.cc#L392-L397

File: include/seastar/rpc/rpc_types.hh [hunk 1]
```diff
@@ -73,6 +73,14 @@ struct client_info {
     typename std::add_const<T>::type& retrieve_auxiliary(const sstring& key) const {
         return const_cast<client_info*>(this)->retrieve_auxiliary<typename std::add_const<T>::type>(key);
     }
+    template <typename T>
+    std::optional<T> retrieve_auxiliary_opt(const sstring& key) const {
```
Comment (gleb-cloudius) [inline include/seastar/rpc/rpc_types.hh:77]: Do we want to copy it? Existing functions return references. In C++ optional reference is a pointer, so my be this is what you want to return here.

Comment (bhalevy) [reply include/seastar/rpc/rpc_types.hh:77]: My initial implementation returned a pointer.
but I wanted to pass it as optional to `netw::msg_addr` constructor.
I can make it an optional in the caller too

Comment (gleb-cloudius) [reply include/seastar/rpc/rpc_types.hh:77]: Yes, the interface should not depend on on use case.

**Comment by bhalevy** [comment]
In v2 ([eda78be](https://github.com/scylladb/seastar/commit/eda78be4ae935a24ba0627cefca11986d9fd166c)):
- return a pointer from retrieve_auxiliary_opt
- added a unit test

**Review by gleb-cloudius — APPROVED** [review]

**Comment by bhalevy** [comment]
@scylladb/seastar-maint please merge

**Pull request closed without merging**